Add LoRa Basics Modem radio drivers and LoRaWAN stack

Salamandar opened the pull request:
This pull request adds support for the Semtech transceivers with the "radio drivers", and for the LoRaWAN stack, both provided by the LoRa Basics Modem project (https://github.com/Lora-net/SWL2001)

This PR adds:
* A new external module, the repository [SWL2001](https://github.com/Lora-net/SWL2001) from Semtech
* Duplicated drivers for LR11xx and SX126x LoRa transceivers -> I postfixed the `compatible` with `-new`, but what's the guideline on that ?
  * The drivers don't expose any API as the LBM repository provides one and the LBM stack uses the "RAL" "Radio Abstraction Layer". We might need to create an ad-hoc API for Zephyr. 
* The LoRaWAN subsystem based on the LBM stack
  * The subsys actually only contains the Kconfig and the source for the "main stack thread". The rest of the code and bindings lives in the `module`.
  * I named the drivers `lora_lbm` and subsys `lorawan_lbm`, to differentiate from the loramac-node implementations.
* Samples: driver samples written from scratch and LBM "examples" ported from the SWL2001 repository.
  * Those ported examples come with the `app_helpers` code in `modules/lora_basics_modem` for portability purpose, but this should be cleaned up.

EDIT: I splitted this PR in half : this PR contains only code that was written with zephyr in mind. The samples that still need porting are in https://github.com/salamandar/zephyr/tree/lbm_samples

zafersn commented:
Hi, @Salamandar, did you have chance to test this with CONFIG_PM?

When I enable it, join request is always fail?

P.S.
I was also working on porting this basic lib to zephyr and finished it just couple of days ago. and just noticed your implementation now :)

zafersn commented:
Hi @Salamandar , I am not quite sure if this is the best place to discuss build failures or not, but just wanted to share these with you, if not, please let me know

 here are errors: when I build your pull request:

Probably you know, but anyway:

disable wdt:
CONFIG_WATCHDOG=n

got error:

zephyr/modules/lora_basics_modem/app_helpers/smtc_hal_zephyr/smtc_hal_watchdog.c:57: undefined reference to `__device_dts_ord_115'

workaround for now.
```
void hal_watchdog_reload(void)
{
#ifdef CONFIG_WATCHDOG
	 wdt_feed(prv_wdt_dev, prv_wdt_channel_id);
#endif
}

```
disabling all region supports:
CONFIG_LORA_BASICS_MODEM_ENABLE_ALL_REGIONS=n

compiling is ok but 

```
*** Booting Zephyr OS build v3.7.0-rc3-124-gabbb8340c2af ***
[00:00:00.000,000] <inf> smtc_modem: Periodical uplink (60 sec) example is starting
[00:00:00.000,000] <inf> smtc_modem: Modem Initialization
[00:00:00.203,000] <inf> smtc_modem: Use soft secure element for cryptographic functionalities
[00:00:00.203,000] <inf> smtc_modem_hal: Opened flash area of size 65536
[00:00:00.203,000] <err> smtc_modem: Invalid Region 0x01
[00:00:00.203,000] <err> smtc_modem: Region (0x1) in nvm context not supported
[00:00:00.203,000] <wrn> smtc_modem: No valid lr1mac context --> Factory reset
[00:00:00.261,000] <inf> smtc_modem: stack_id 0
[00:00:00.261,000] <inf> smtc_modem: DevNonce = 0
[00:00:00.261,000] <inf> smtc_modem: JoinNonce = 0xff ff ff, NetID = 0xff ff ff
[00:00:00.261,000] <inf> smtc_modem: Region = CN470
[00:00:00.261,000] <inf> smtc_modem: LoRaWAN Certification is disabled on stack 0
[00:00:00.261,000] <inf> smtc_modem: lorawan_alcsync_services_init task_id 7, service_id 0, CURRENT_STACK:0
[00:00:00.261,000] <wrn> smtc_modem: stream_services_init
[00:00:00.261,000] <wrn> smtc_modem: lfu_services_init
[00:00:00.290,000] <inf> smtc_modem: Modem event callback
[00:00:00.290,000] <inf> smtc_modem: Event received: RESET
[00:00:00.290,000] <err> smtc_modem: Invalid Region 0x01
[00:00:00.291,000] <err> smtc_modem: In WEST_TOPDIR/wsc/periodical_uplink_semtec/src/main_periodical_uplink.c - modem_event_callback (line 373): SMTC_MODEM_RC_INVALID
[00:00:00.291,000] <inf> smtc_modem: smtc_modem_join_network
[00:00:00.291,000] <inf> smtc_modem: Start a new join sequence in 3 seconds on stack 0
[00:00:00.291,000] <inf> smtc_modem: Starting loop...
[00:00:00.291,000] <inf> smtc_modem: Sleeping for 3000ms
[00:00:03.291,000] <inf> smtc_modem: Sleeping for 0ms
[00:00:03.291,000] <inf> smtc_modem: Sleeping for 10000ms
[00:00:06.859,000] <inf> smtc_modem: modem_tpm_radio_busy_lbt
[00:00:06.859,000] <inf> smtc_modem: Sleeping for 10000ms
```

Ill share the quick workaround solution  shortly

Salamandar committed:
west.yml: Add project LoRa Basics Modem (to be forked in zephyr github org)

This project will be used in the next commits.

Signed-off-by: Félix Piédallu <felix@piedallu.me>

Salamandar committed:
modules: Add lora_basics_modem module

This module provides:
* The cmake files for LoRa Basics Modem
* The "HAL" described here:
  https://github.com/Lora-net/SWL2001/blob/master/lbm_lib/PORTING_GUIDE.md
* Reimplementations of the logging functions in smtc_modem_hal_dbg_trace.h

Signed-off-by: Félix Piédallu <felix@piedallu.me>

Salamandar committed:
drivers, dts: Add drivers for Semtech LoRa transceivers LR11xx, SX126x

Those drivers are actually device tree bindings, and HAL/BSP as described
in https://github.com/Lora-net/SWL2001/blob/master/lbm_lib/PORTING_GUIDE.md

Signed-off-by: Félix Piédallu <felix@piedallu.me>

Salamandar committed:
lorawan_lbm: Add LoRa Basics Modem LoRaWAN stack

This stack is based on the lora_basics_modem external module.

It provides configuration for LBM as well as a standalone
"main" LBM thread that needs to be manually started.

Signed-off-by: Félix Piédallu <felix@piedallu.me>

Salamandar committed:
boards: shields: Add configuration for LoRa LR11xx and SX126x shields

Those shields are sold by Semtech and are arduino compatible.

Signed-off-by: Félix Piédallu <felix@piedallu.me>

Salamandar committed:
MAINTAINERS: Add LoRa Basics Modem files to the Maintainers.yml

Add the module, subsys, drivers for LoRa Basics Modem.
Add Salamandar (Félix Piédallu) to its maintainers.

Signed-off-by: Félix Piédallu <felix@piedallu.me>

Salamandar commented:
> Hi @Salamandar I’m not sure if this is the best place to discuss build failures, but I wanted to share these with you. If there’s a more appropriate forum or person for this, please let me know.
> disable wdt: CONFIG_WATCHDOG=n

Ah yeah, that's "expected". The code in the samples and `lora_basics_modem/app_helpers` mostly is copy-pasta from the official samples and example codes, and it's not really "portable" nor configurable yet. I'll work towards a cleaner code in this PR.

ndrs-pst commented:
@KajbaM, forgive the ping!
Since IRNAS has implemented similar functionality with the Lora Basic Modem for Zephyr, 
it would be fantastic to get insights on this PR.
Your experience with [SWL2001-Zephyr](https://github.com/IRNAS/SWL2001-Zephyr) would be incredibly helpful. Thank you!

TjazVracko commented:
@ndrs-pst I work at IRNAS and will submit a review. This is more or less a continuation of our work by Semtech, so we might have some additional insight.

Hopefully, this gets merged, so we can all benefit :)

TjazVracko commented on subsys/lorawan_lbm/lbm_main_thread.c:
```diff
@@ -0,0 +1,55 @@
+		if (smtc_modem_is_irq_flag_pending()) {
+			continue;
+		}
+		LOG_INF("Sleeping for %dms", sleep_time_ms);
```
```suggestion
		LOG_DBG("Sleeping for at most %dms", sleep_time_ms);
```

So that logs are less spammy.

TjazVracko commented on subsys/lorawan_lbm/Kconfig:
```diff
@@ -0,0 +1,341 @@
+# Miscellaneous options
+#-----------------------------------------------------------------------------
+
+config LORA_BASICS_MODEM_NUMBER_OF_STACKS
+	int "Number of LoRa Basics Modem stacks (only 1 is supported for now)"
+	default 1
+	help
+	 Number of stacks to be used by the modem library.
```
Since 1 is the only valid value, this should not be user-assignable:

```suggestion
config LORA_BASICS_MODEM_NUMBER_OF_STACKS
	int
	default 1  # only 1 is supported for now
	help
	 Number of stacks to be used by the modem library.
```

TjazVracko commented on samples/drivers/lora_lbm/ping_pong/src/main_ping_pong.c:
```diff
@@ -0,0 +1,250 @@
+{
+	LOG_INF("===== SX126x TX CW example =====");
+
+	apps_common_lr11xx_system_init(context);
```
You use `apps_common_` functions here, but there is no `apps_common` module.

Salamandar replied:
Ah yes, indeed, that's part of the samples that were copy-pasted from SWL2001, but should be reworked for a cleaner integration in Zephyr.

TjazVracko reviewed (commented):
I left some general comments and suggestions.

Other maintainers should leave comments about more Zephyr-specific topics.

---------------------------
If this is merged, you will then have 2 LoRaWAN stacks in Zephyr - LBM and LoRaMAC-Node.
Since LBM is the new and supported one, could `LoRaMAC-Node` be removed? It should be trivial to implement the Zephyr LoRa and LoRaWAN APIs on top of LBM.
Thoughts?

Salamandar commented on west.yml:
```diff
@@ -278,6 +280,11 @@ manifest:
+    - name: lora_basics_modem
+      remote: lora-net
+      repo-path: SWL2001
+      revision: feature/zephyr
```
Yes, indeed, we would need to point on a specific tag. I left it as is to be able to iterate on the reviews :) Also, Semtech will release a new version of SWL2001 that integrates the patches present on this zephyr-specific branch.

Salamandar commented on boards/shields/semtech_lr11xxmb1xxs/semtech_lr11xxmb1xxs_common.dtsi:
```diff
@@ -0,0 +1,170 @@
+	};
+};
+
+// FIXME: Is there a way to reassign the node label arduino_spi from spi2 to spi3 ?
+// spi2 draws way less current than spi3.
```
Actually that is related to the nRF52840 hardware specs, see https://devzone.nordicsemi.com/f/nordic-q-a/104177/high-power-consumption-when-using-spim3

Salamandar commented on modules/lora_basics_modem/common.cmake:
```diff
@@ -0,0 +1,394 @@
+# ALCSync
+#-----------------------------------------------------------------------------
+
+zephyr_compile_definitions_ifdef(CONFIG_LORA_BASICS_MODEM_ALC_SYNC
```
Yes, that is actually something I wanted to check (there are some `ifdef`s in the public header, so that's something to be done with caution).
Thank you for the reminder, I had forgot about that.

Salamandar replied:
See 29821ec2d1758c445c3b0cba9a914db852fc95cc and 2fd9f86d997abf38f31989cdcea8c3e74fb346b9.

(i'll squash the commits when all issues are fixed)

Salamandar commented on modules/lora_basics_modem/lr11xx.cmake:
```diff
@@ -0,0 +1,64 @@
+  ${LBM_LR11XX_DIR}/lr11xx_lr_fhss.c
+)
+
+# LR1121 is the only one not supporting GNSS and WiFi
+zephyr_library_sources_ifndef(CONFIG_DT_HAS_SEMTECH_LR1121_ENABLED
+  ${LBM_LR11XX_DIR}/lr11xx_wifi.c
+  ${LBM_LR11XX_DIR}/lr11xx_gnss.c
+)
```
It does not unfortunately, and LBM does not support being instantiated multiple times either. That's why I took the freedom to do that. Also LBM doesn't have the capability to do runtime-capability check so *not* including the code is the only safe way to disable it for specific radios.

Salamandar commented on modules/lora_basics_modem/CMakeLists.txt:
```diff
@@ -0,0 +1,48 @@
+  set(LORA_BASICS_MODEM_DIR ${ZEPHYR_CURRENT_MODULE_DIR})
+  set(LORA_BASICS_MODEM_LIB_DIR ${LORA_BASICS_MODEM_DIR}/lbm_lib)
+  set(LBM_SMTC_MODEM_CORE_DIR ${LORA_BASICS_MODEM_LIB_DIR}/smtc_modem_core)
+  set(LBM_RADIO_DRIVERS_DIR ${LORA_BASICS_MODEM_LIB_DIR}/smtc_modem_core/radio_drivers)
```
Actually I'll keep `LORA_BASICS_MODEM_DIR` for consistency with the rest of the modules, but the other variable makes complete sense.

Salamandar commented on drivers/lora_lbm/lr11xx/lr11xx_board.c:
```diff
@@ -0,0 +1,317 @@
+				 (COND_CODE_1(DT_NODE_HAS_COMPAT(node_id, semtech_lr1121),         \
+					      (LR11XX_SYSTEM_VERSION_TYPE_LR1121), (/**/))))))
+
+/* LR11xx does not support external TCXO control */
+#define LR11XX_CFG_TCXO(node_id)                                                                   \
```
I should make it clearer I think. By "external control" I mean "TCXO control handled by the software on the MCU". SX126x support this, but not LR11xx.

Salamandar commented on dts/bindings/lora_lbm/semtech,lr11xx-common.yaml:
```diff
@@ -0,0 +1,253 @@
+
+      Can be reconfigured at runtime via radio_utilities_set_tx_power_offset.
+
+  tx-power-cfg-lf-lp:
```
> Would it be possible to have a default set of these defined separately in some files, and then include them in the DTS?

Ah yes, that would be a nice idea, I like that.

erwango commented:
@Salamandar Is that still actually a Draft work ?
In that case, please use the "Convert to draft" button below the reviewers pane, otherwise, please update the title.
Aim is to better understand where you're at

Salamandar commented on modules/lora_basics_modem/smtc_modem_hal/smtc_modem_hal_init.h:
```diff
@@ -0,0 +1,182 @@
+ * <transceiver>_board_attach_interrupt(). To resume usage of defout irq callbacks:
+ * prv_transceiver_event_cb(), this function must be called, before resuming usage of
+ * smtc modem by calling smtc_modem_resume_after_user_radio_access(), otherwise
+ * event handling will not work.
```
This was supposed to be dropped. Fixed.

Salamandar commented:
> @Salamandar Is that still actually a Draft work ?

It is still a work in progress but, indeed, not a draft anymore.

I expect a lot of comments, and requests for changes, that's a bit why I named it a "draft". But you're right, that is misleading on the state of the code.

Salamandar commented:
FYI the compliance checks fail on something where clang-format and Checkpatch disagree, see https://github.com/zephyrproject-rtos/zephyr/pull/83374.

martinjaeger commented:
As a general comment: We try to avoid having multiple implementation for the same thing in the Zephyr tree. This means that the new LoRa Basics Modem will finally have to replace the "old" LoRaMAC-node stack. If it's not feasible to replace the entire stack at once, we can have a transition phase with both options in parallel (implementation selectable via Kconfig). But we should ideally try to use the same user-facing API for both.

I suggest to split this PR into several smaller chunks which make it easier to review:

- If that makes sense, we could update just the LoRa driver first and the entire LoRaWAN stack in a follow-up PR?
- New boards/shields/samples should be added in a follow-up PR once the driver and LoRaWAN stack updates are merged.

Existing devicetree bindings should be reused (and changed/extended, if required).

Do you think we can (mostly) re-use the existing APIs in `include/zephyr/drivers/lora.h` and `include/zephyr/lorawan/lorawan.h` with the new stack?

wietsedv commented:
Semtech appears to have written an [out-of-tree LBM Zephyr driver](https://github.com/Lora-net/LBM_Zephyr/) themselves. It would probably be a good idea to compare these two implementations.

kartben commented:
> Semtech appears to have written an [out-of-tree LBM Zephyr driver](https://github.com/Lora-net/LBM_Zephyr/) themselves. It would probably be a good idea to compare these two implementations.

well this PR is from Semtech, so basically the same code (or a subset of it, at least) :)

JordanYates commented:
I am planning to start looking at this in the near future.
Since this PR is quite large, I think a better process is going to be building up support in-tree in smaller chunks, starting with just the basic LoRa API (packet send/receive).
This will let us get the infrastructure (repo fork, build system etc) in place without an overwhelming diff.
We can then look at the LoRaWAN integration, new boards, etc in follow-on PRs.

martinjaeger commented:
@JordanYates FYI: Unfortunately, Semtech decided not to continue funding the upstreaming process of the LoRa Basics Modem stack to Zephyr and leave it to the community. Before you spend much time into review the PR, we should check with @Salamandar if he will be able to continue working on this PR.

fmoessbauer commented:
As I just stumbled upon the shortcomings of the `loramac-node` driver, I'm scratching my head around the following questions:

- Is the sx1276 chip supported (can't find any reference in this PR, but upstream [supports it](https://github.com/Lora-net/SWL2001/tree/9a14f6772c1d6e303bacb2d594c8063bb804b6ee/lbm_lib/smtc_modem_core/radio_drivers/sx127x_driver)). The [Lilygo T-Beam](https://docs.zephyrproject.org/latest/boards/lilygo/ttgo_tbeam/doc/index.html) has it which is widely used for LoRa HAM operations.
- Is it possible to set the LoRa sync_word to a user-provided value? The RAL [supports this](https://github.com/Lora-net/SWL2001/blob/9a14f6772c1d6e303bacb2d594c8063bb804b6ee/lbm_lib/smtc_modem_core/smtc_ral/src/ral.h#L705), but it is crucial to also expose this API in Zephyr. This is currently a showstopper of porting Meshcom or Meshtastic to Zephyr. For details, see [this discussion](https://github.com/zephyrproject-rtos/zephyr/discussions/78995).

JordanYates commented:
> Is the sx1276 chip supported

It is not supported in this PR, but I have it working to some extent with LBM, which will be part of my initial transition PR.

> Is it possible to set the LoRa sync_word to a user-provided value? 

Not with the current Zephyr LoRa API. There is nothing stopping us from extending the API to support this, although I am a bit confused since the RAL abstraction only takes a `uint8_t`, but the `loramac-node` defines differ in size (and value) between the sx126x and sx127x drivers.
```
// SX1272.h
#define LORA_MAC_PRIVATE_SYNCWORD                   0x12
// SX126x.h
#define LORA_MAC_PRIVATE_SYNCWORD                   0x1424
```

fmoessbauer commented:
> I am a bit confused since the RAL abstraction only takes a uint8_t, but the loramac-node defines differ in size (and value) between the sx126x and sx127x drivers.

The LoRa sync words are complicated. It's not simply that nodes can communicate if the sync words match. Instead, a transmission only reliably works if certain conditions between the sender and receiver syncword match. That's why there are different syncwords between SX1272 (2-byte) and SX126x (4-byte) for the same network.

The chosen values also have an effect on the likelyhood of a successful transmission (it's not just a deterministic criterion). Further, not all sync words are valid at all. Unfortunately all that is undocumented territory. The probably best analysis is found in this blog post: https://blog.classycode.com/lora-sync-word-compatibility-between-sx127x-and-sx126x-460324d1787a

**TL;DR:**
This technically means, we need two sync words (one 2-byte, one 4-byte) in the abstraction layer, as we don't know the hardware below. The driver can then just pick the corresponding one. As these sync words cannot easily be derived from each other, both must be user-provided. This is also how Meshtastic and Meshcom currently handle it: Provide both and `#ifdef` based on chip (which would be done in the driver part then).

fmoessbauer commented:
I just stumbled upon a derivation function for the sync word in RadioLib: https://github.com/jgromes/RadioLib/blob/c570b44dc35f3d2cddf326926b17b1ea91e32328/src/modules/SX128x/SX128x.cpp#L1020

```
uint8_t data[2] = {(uint8_t)((syncWord & 0xF0) | ((controlBits & 0xF0) >> 4)),
                   (uint8_t)(((syncWord & 0x0F) << 4) | (controlBits & 0x0F))};
```

This means the sync word is just 2 bytes and then another 2 bytes of control bits are mixed in, which are simply `0x44` (per default). If this is just a coincidence or a rule remains unclear. At least in case of Meshcom this is not true, as they use `0x2b` and `0x2b24` (but it is even unclear if this works due to coincidence or by design).

JordanYates commented:
The equation you came across looks quite similar to the conversion in LoRa Basics Modem to convert the `uint8_t` in the RAL API to the 2 byte sync word for the SX126x

https://github.com/Lora-net/SWL2001/blob/9a14f6772c1d6e303bacb2d594c8063bb804b6ee/lbm_lib/smtc_modem_core/radio_drivers/sx126x_driver/src/sx126x.c#L1310-L1325

fmoessbauer commented:
Indeed, but there they just read the sync word from the modem (`SX126X_REG_LR_SYNCWORD`) to get the control bits and then mix the 2-byte sync word in. Is that even a valid call-chain according to the datasheet of the `sx126x`, in case no sync word was set before?

If there only would be documentation...
